storage/file_storage.go

```diff
@@ -0,0 +1,179 @@
+package storage
+
+import (
+	"encoding/json"
+	"os"
+	"sync"
+	"time"
+)
+
+// FileStorage represents the data structure for JSON serialization
+type FileStorage struct {
+	Gauges   map[string]float64 `json:"gauges"`
+	Counters map[string]int64   `json:"counters"`
+}
+
+// FileManager handles file operations for metrics storage
+type FileManager struct {
+	filePath string
+	mu       sync.RWMutex
+}
+
+// NewFileManager creates a new file manager
+func NewFileManager(filePath string) *FileManager {
+	return &FileManager{
+		filePath: filePath,
+	}
+}
+
+// SaveToFile saves the current metrics to file
+func (fm *FileManager) SaveToFile(storage Storage) error {
```

Comment:
Я бы предложил передавать ссылку на  `Storage` в фабричный метод и хранит в инстансе структуры.
```suggestion
// FileManager handles file operations for metrics storage
type FileManager struct {
	filePath string
	storage Storage
	mu       sync.RWMutex
}

// NewFileManager creates a new file manager
func NewFileManager(filePath string, storage Storage) *FileManager {
	return &FileManager{
		filePath: filePath,
		storage: storage,
	}
}

// SaveToFile saves the current metrics to file
func (fm *FileManager) SaveToFile() error {
```